### PR TITLE
feat: chart yaxis adaptive

### LINKED ISSUE #6210: [Enhancement] Ability to set the y-axis adaptive based on the data in Chart widget
The chart widget today doesn't have any way to make the y-axis adaptive based on the data.
A toggle that makes the y-axis adaptive would solve this.

https://www.fusioncharts.com/dev/chart-guide/chart-configurations/chart-limits#set-y-axis-minimum-value

### PR DESCRIPTION
**Pull request by techbhavin**
## Description

> Ability to set the y-axis adaptive based on the data in Chart widget

Fixes #6210 

## Type of change

- Bugfix (non-breaking change which fixes an issue)

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ]  I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ]  I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

## Test coverage results :test_tube:
<details><summary>:white_circle: Total coverage has not changed</summary>

    // Code coverage diff between base branch:release and head branch: feature/chart-yaxis-adaptive 
Status | File | % Stmts | % Branch | % Funcs | % Lines 
 -----|-----|---------|----------|---------|------ 
 :green_circle: | total | 54.94 **(0)** | 36.8 **(-0.01)** | 33.69 **(0.01)** | 55.48 **(0)**
 :green_circle: | app/client/src/components/designSystems/appsmith/ChartComponent.tsx | 49.41 **(0)** | 22.11 **(0.13)** | 50 **(0)** | 47.47 **(0)**
 :red_circle: | app/client/src/utils/autocomplete/TernServer.ts | 50.47 **(-0.24)** | 37.72 **(-0.88)** | 36.21 **(0)** | 55.08 **(-0.27)**</details>

### CODE REVIEW

**Review by somangshu — DISMISSED** [review]
Approving to run  tests

File: app/client/src/widgets/ChartWidget/propertyConfig.ts [hunk 1]
```diff
@@ -232,6 +232,15 @@ export default [
         isTriggerProperty: false,
         hidden: (x: any) => x.chartType === "CUSTOM_FUSION_CHART",
       },
+      {
+        propertyName: "setAdaptiveYMin",
+        label: "Y-axis Minimum",
+        helpText: "Toggle y-axis adaptive min",
```
Comment (somangshu) [inline app/client/src/widgets/ChartWidget/propertyConfig.ts:238]: Change to:
> Define the minimum scale for Y axis

File: app/client/src/components/designSystems/appsmith/ChartComponent.tsx [hunk 2]
```diff
@@ -239,6 +240,7 @@ class ChartComponent extends React.Component<ChartComponentProps> {
       captionAlignment: "left",
       captionHorizontalPadding: 10,
       alignCaptionWithCanvas: 0,
+      setAdaptiveYMin: this.props.setAdaptiveYMin ? "1" : "0",
```
Comment (somangshu) [inline app/client/src/components/designSystems/appsmith/ChartComponent.tsx:243]: We should allow the dev to set the value here, The min could be 50 as well

File: app/client/src/components/designSystems/appsmith/ChartComponent.tsx [hunk 3]
```diff
@@ -51,6 +51,7 @@ export interface ChartComponentProps {
   widgetId: string;
   isVisible?: boolean;
   allowHorizontalScroll: boolean;
+  setAdaptiveYMin: boolean;
```
Comment (somangshu) [inline app/client/src/components/designSystems/appsmith/ChartComponent.tsx:54]: This should be an input field of type number

**Comment by somangshu** [comment]
@techbhavin from the deploy preview:
- This does not change the y axis scale in the chart widget
- The property should also be available for the custom chart config.

**Comment by somangshu** [comment]
@techbhavin the functionality still done not work with any of the charts. Setting Y-Min does not change the scale

- Additionally the field should have a default of 0
- We need to find a way to make this work for the custom chart out of the box. If adding it to the config works then thats fine

@jsartisan can you help out here

**Review by jsartisan — APPROVED** [review]

**Review by somangshu — DISMISSED** [review]
Approving to run tests

**Comment by jsartisan** [comment]
@techbhavin The current implementation is not ideal. I just discussed this with @somangshu.

What we will be needing is two fields in the property pane.

1. `Y-axis adaptive` -> this will be a toggle. This field will set the `setAdaptiveYMin` in fusioncharts to 1.
2. `Y-axis minimum` -> this will be an input text box. This filed will set the `yAxisMinValue` in fusioncharts.

Now if `Y-axis adaptive`  is on, `Y-axis minimum` will be hidden.
`Y-axis adaptive` being true means the user wants to set the minimum/maximum automatically based on the data.
`Y-axis minimum` means the user wants to set the minimum value explicitly.

@somangshu @techbhavin Let me know if there any help required from my side.

**Review by jsartisan — DISMISSED** [review]

**Comment by jsartisan** [comment]
LGTM!

**Comment by jsartisan** [comment]
@techbhavin just one thing, can you update the label of that toggle from "Y-axis minimum" to "Y-axis adaptive" ?

**Review by jsartisan — DISMISSED** [review]

**Comment by somangshu** [comment]
The failing build is due to the newly added D&D. @marks0351 will be looking into this

**Review by somangshu — DISMISSED** [review]

**Comment by YogeshJayaseelan** [comment]
@techbhavin @somangshu - Y Axis adaptive option is applicable for Line chart where as for bar chart X axis values are getting updated and also for pie chart Y axis adaptive is not required

https://www.loom.com/share/37ceb53a99304fa4ae65da396d33c8d0

**Comment by somangshu** [comment]
@YogeshJayaseelan Leaving it as it is for all chart types is fine. What I feel confused about is that the x / y axis are adaptive and might not be expected, In that case we need to rethink the naming convention for this, Maybe we should call it `Adaptive Axis`. 

@jsartisan @techbhavin what are your thoughts?

**Comment by somangshu** [comment]
@techbhavin as per the above, Please change the name of the property to `Adaptive Axis`

**Review by jsartisan — APPROVED** [review]

**Review by somangshu — APPROVED** [review]

**Comment by somangshu** [comment]
@techbhavin please merge the latest release here

**Comment by YogeshJayaseelan** [comment]
@techbhavin Verified the fix for Adaptive Axis and it is working as expected for Line, Bar, column, Area charts 

Issues:
2. Every-time when adaptive axis is enabled or disabled for custom chart the chart reloads each and every time - @somangshu need you confirmation here whether is it an expected behaviour or we should fix the reload 

https://www.loom.com/share/410d093d56dd4ca0a87919bfbad5317f

3.The tool tip message for 'Adaptive axis' option reads as 'Define the minimum scale for Y axis' and it should be renamed to 'Define the minimum scale for X/Y axis'

Screenshot 2021-09-01 at 9 29 35 AM

**Review by jsartisan — DISMISSED** [review]

**Review by jsartisan — APPROVED** [review]

**Comment by YogeshJayaseelan** [comment]
@techbhavin verified issue 3 and it is working as expected now and issue 2 seems to be normal behaviour and moving this to done state

**Pull request merged**